neqo-transport/src/connection/tests/ecn.rs

```diff
@@ -0,0 +1,512 @@
+    assert_ecn_enabled(before); // Too few packets sent before migration to conclude ECN validation.
+    assert_ecn_enabled(after); // Too few packets sent after migration to conclude ECN validation.
+}
+
+#[test]
+fn ecn_migration_remark_drop_nodata() {
+    let (before, after) = migration_with_modifiers(remark(), drop(), 0);
+    assert_ecn_enabled(before); // Too few packets sent before migration to conclude ECN validation.
+    assert_ecn_enabled(after); // Migration failed, still too few packets to conclude ECN
+                               // validation.
+}
+
+#[test]
+fn ecn_migration_remark_noop_data() {
+    let (before, after) = migration_with_modifiers(remark(), noop(), ECN_TEST_COUNT);
+    assert_ecn_disabled(before); // ECN validation fails before migration due to remarking.
+    assert_ecn_enabled(after); // ECN validation succeeds after migration.
+}
+
+#[test]
+fn ecn_migration_remark_bleach_data() {
+    let (before, after) = migration_with_modifiers(remark(), bleach(), ECN_TEST_COUNT);
+    assert_ecn_disabled(before); // ECN validation fails before migration due to remarking.
+    assert_ecn_disabled(after); // ECN validation fails after migration due to bleaching.
+}
+
+#[test]
+fn ecn_migration_remark_remark_data() {
+    let (before, after) = migration_with_modifiers(remark(), remark(), ECN_TEST_COUNT);
+    assert_ecn_disabled(before); // ECN validation fails before migration due to remarking.
+    assert_ecn_disabled(after); // ECN validation fails after migration due to remarking.
+}
+
+#[test]
+fn ecn_migration_remark_ce_data() {
+    let (before, after) = migration_with_modifiers(remark(), ce(), ECN_TEST_COUNT);
+    assert_ecn_disabled(before); // ECN validation fails before migration due to remarking.
+    assert_ecn_enabled(after); // ECN validation concludes after migration, despite all CE marks.
+}
+
+#[test]
+fn ecn_migration_remark_drop_data() {
+    let (before, after) = migration_with_modifiers(remark(), drop(), ECN_TEST_COUNT);
+    assert_ecn_disabled(before); // ECN validation fails before migration due to remarking.
+    assert_ecn_disabled(after); // Migration failed, ECN on original path is still disabled.
+}
+
+#[test]
+fn ecn_migration_ce_noop_nodata() {
+    let (before, after) = migration_with_modifiers(ce(), noop(), 0);
+    assert_ecn_enabled(before); // Too few packets sent before migration to conclude ECN validation.
+    assert_ecn_enabled(after); // Too few packets sent after migration to conclude ECN validation.
+}
+
+#[test]
+fn ecn_migration_ce_bleach_nodata() {
+    let (before, after) = migration_with_modifiers(ce(), bleach(), 0);
+    assert_ecn_enabled(before); // Too few packets sent before migration to conclude ECN validation.
+    assert_ecn_enabled(after); // Too few packets sent after migration to conclude ECN validation.
+}
+
+#[test]
+fn ecn_migration_ce_remark_nodata() {
+    let (before, after) = migration_with_modifiers(ce(), remark(), 0);
```

Comment:
If - when the burst is 0 - all the connections succeed with ECN enabled before and after, why not make a single test that loops through all of the functions?  I know that if you are testing whether migration occurred, you will need to check for noop, but that's just one extra check.

Right here, you have a lot of copy-pasted code.

Reply:
But on second thought this special case makes sense to consolidate.

Reply:
```rust
#[test]
fn ecn_migration_zero_burst_all_cases() {
    let modifiers = [noop(), bleach(), remark(), ce(), drop()];
    for orig_path_modifier in modifiers.iter() {
        for new_path_modifier in modifiers.iter() {
            let (before, after) =
                migration_with_modifiers(orig_path_modifier, new_path_modifier, 0);
            // Too few packets sent before and after migration to conclude ECN validation.
            assert_ecn_enabled(before);
            assert_ecn_enabled(after);
        }
    }
}
```
@martinthomson any idea how to work around this error?
```
error[E0308]: mismatched types
   --> neqo-transport/src/connection/tests/ecn.rs:227:30
    |
38  | fn noop() -> impl DatagramModifier {
    |              --------------------- the expected opaque type
...
42  | fn bleach() -> impl DatagramModifier {
    |                --------------------- the found opaque type
...
227 |     let modifiers = [noop(), bleach(), remark(), ce(), drop()];
    |                              ^^^^^^^^ expected opaque type, found a different opaque type
    |
    = note: expected opaque type `impl connection::tests::DatagramModifier` (opaque type at <neqo-transport/src/connection/tests/ecn.rs:38:14>)
               found opaque type `impl connection::tests::DatagramModifier` (opaque type at <neqo-transport/src/connection/tests/ecn.rs:42:16>)
    = note: distinct uses of `impl Trait` result in different opaque types
```

Reply:
This works
```rust
    let modifiers = [Some, |d| Some(set_tos(d, IpTosEcn::NotEct))];
```
but it obviously duplicates what `noop()` and `bleach()`, etc., are doing...

Reply:
Nevermind, figure it out.